gnosis/verification.k

```diff
@@ -6,21 +6,35 @@ module VERIFICATION
     imports EDSL
     imports LEMMAS
 
+  // ########################
+  // Memory Usage
+  // ########################
+
+    rule #memoryUsageUpdate(#memoryUsageUpdate(MU, START0, WIDTH0), START1, WIDTH1) => #memoryUsageUpdate(MU, START0, WIDTH0)
+      requires START1 +Int WIDTH1 <Int START0 +Int WIDTH0
+
+    rule #memoryUsageUpdate(#memoryUsageUpdate(MU, START0, WIDTH0), START1, WIDTH1) => #memoryUsageUpdate(MU, START1, WIDTH1)
+      requires START0 +Int WIDTH0 <=Int START1 +Int WIDTH1
+
+    rule 0 <=Int #memoryUsageUpdate(MU, START, WIDTH) => true
+
+    rule pow256 <=Int #memoryUsageUpdate(MU, START, WIDTH) => false  requires START +Int WIDTH <Int 3705346855594118253554271520278013051304639509300498049262642688253220148477921
+    rule #memoryUsageUpdate(MU, START, WIDTH) <Int  pow256 => true   requires START +Int WIDTH <Int 3705346855594118253554271520278013051304639509300498049262642688253220148477921
+
+
   // ########################
   // Gas
   // ########################
 
-    // Temporary solution to avoid reasoning about gas
-    syntax Int ::= "#infinity" [function]
- // -------------------------------------
-    rule #infinity +Int _ => #infinity
-    rule #infinity -Int _ => #infinity
-    rule A <=Int #infinity => true
-      requires #getKLabelString(A) =/=String "#infinity_VERIFICATION"
-    rule #infinity <Int A => false
-      requires #getKLabelString(A) =/=String "#infinity_VERIFICATION"
+    syntax Int ::= #gas ( Int , Int ) [function]
+ // --------------------------------------------
+    rule #gas(NONMEM, MEM) -Int #symCmem(MEM') => #gas(NONMEM, MEM +Int MEM')
```

Comment:
why `MEM +Int MEM'` not `MEM -Int MEM'` ?

Reply:
ah, #gas(...) is the accumulated amount of deducted gas?

Reply:
We are accumulating all the gas during the execution, right?  MEM' ==K Cmem(New) -Cmem(OLD) is the gas cost. So we accumulate the cost in the mem part of #gas .

Reply:
is this #gas(...) stored in <gas> cell, which denotes the remaining gas, right?  then, #gas(...) should take three arguments, say #gas(ORIGIN, NONMEM, MEM), which denotes ORIGIN - NONMEM - MEM ?

Reply:
you're correct. But what value do we put for ORIGIN? a fresh variable?

Reply:
like `#gas(#infinity, 0, 0)` ?

Reply:
what i mean by `#gas(#infinity, 0, 0)`, is the LHS of the spec.

for the ORIGIN, i think we need to introduce `#infinity` again, but without any lemma directly over it.